src/settings/handlers/AbstractLocalStorageSettingsHandler.ts

```diff
@@ -0,0 +1,82 @@
+/*
+Copyright 2019 - 2022 The Matrix.org Foundation C.I.C.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import SettingsHandler from "./SettingsHandler";
+
+/**
+ * Abstract settings handler wrapping around localStorage making getValue calls cheaper
+ * by caching the values and listening for localStorage updates from other tabs.
+ */
+export default abstract class AbstractLocalStorageSettingsHandler extends SettingsHandler {
+    private itemCache = new Map<string, any>();
+    private objectCache = new Map<string, object>();
+
+    protected constructor() {
+        super();
+
+        // Listen for storage changes from other tabs to bust the cache
+        window.addEventListener("storage", (e: StorageEvent) => {
+            this.itemCache.delete(e.key);
+            this.objectCache.delete(e.key);
```

Comment:
It looks like there's a special case where `e.key` is `null` when `clear()` is called.
https://developer.mozilla.org/en-US/docs/Web/API/StorageEvent

In that scenario I'd imagine we would want to clear our entire local cache too

Reply:
Good call!